### PR TITLE
[#196] Network rate limit for backup

### LINKED ISSUE #196: Rate limit: Network
Provide a rate limit (global and server) to limit network traffic.

This is for backup

### PR DESCRIPTION
**Pull request by shikharish**
Closes #196

### CODE REVIEW

File: src/libpgmoneta/configuration.c [hunk 1]
```diff
@@ -1183,6 +1184,20 @@ pgmoneta_read_configuration(void* shm, char* filename)
                      unknown = true;
                   }
                }
+               else if (!strcmp(key, "network_max_rate"))
+               {
+                  if (!strcmp(section, "pgmoneta"))
```
Comment (jesperpedersen) [inline src/libpgmoneta/configuration.c:1189]: We should do this per server too

**Review by jesperpedersen — CHANGES_REQUESTED** [review]

**Comment by jesperpedersen** [comment]
Need documentation as well

File: doc/CONFIGURATION.md [hunk 2]
```diff
@@ -72,6 +72,8 @@ See a [sample](./etc/pgmoneta.conf) configuration for running `pgmoneta` on `loc
...
 | update_process_title | `verbose` | String | No | The behavior for updating the operating system process title. Allowed settings are: `never` (or `off`), does not update the process title; `strict` to set the process title without overriding the existing initial process title length; `minimal` to set the process title to the base description; `verbose` (or `full`) to set the process title to the full description. Please note that `strict` and `minimal` are honored only on those systems that do not provide a native way to set the process title (e.g., Linux). On other systems, there is no difference between `strict` and `minimal` and the assumed behaviour is `minimal` even if `strict` is used. `never` and `verbose` are always honored, on every system. On Linux systems the process title is always trimmed to 255 characters, while on system that provide a natve way to set the process title it can be longer. |
+| network_max_rate | 0 | Int | No | The number of bytes of tokens added every one second to limit the netowrk backup rate. 0 means no limit.|
```
Comment (jesperpedersen) [inline doc/CONFIGURATION.md:75]: Move next to `backup_max_rate`

File: doc/CONFIGURATION.md [hunk 3]
```diff
@@ -90,6 +92,8 @@ See a [sample](./etc/pgmoneta.conf) configuration for running `pgmoneta` on `loc
 | tls_cert_file | | String | No | Certificate file for TLS. This file must be owned by either the user running pgmoneta or root. |
 | tls_key_file | | String | No | Private key file for TLS. This file must be owned by either the user running pgmoneta or root. Additionally permissions must be at least `0640` when owned by root or `0600` otherwise. |
 | tls_ca_file | | String | No | Certificate Authority (CA) file for TLS. This file must be owned by either the user running pgmoneta or root.  |
+| network_max_rate | 0 | Int | No | The number of bytes of tokens added every one second to limit the netowrk backup rate. 0 means no limit or uses global configuration.|
```
Comment (jesperpedersen) [inline doc/CONFIGURATION.md:95]: Same - but `-1` is default

File: src/include/pgmoneta.h [hunk 4]
```diff
@@ -228,6 +228,7 @@ struct server
    char tls_key_file[MISC_LENGTH];    /**< TLS key path */
    char tls_ca_file[MISC_LENGTH];     /**< TLS CA certificate path */
    int workers;                       /**< The number of workers */
+   unsigned long network_max_rate;    /**< Number of bytes of tokens added every one second to limit the netowrk backup rate */
```
Comment (jesperpedersen) [inline src/include/pgmoneta.h:231]: Move next to `backup_max_rate` and I changed that data type

File: src/include/pgmoneta.h [hunk 5]
```diff
@@ -355,6 +356,7 @@ struct configuration
    int number_of_admins;         /**< The number of admins */
 
    unsigned long backup_max_rate; /**< Number of tokens added to the bucket with each replenishment for backup. */
+   unsigned long network_max_rate;    /**< Number of bytes of tokens added every one second to limit the netowrk backup rate */
```
Comment (jesperpedersen) [inline src/include/pgmoneta.h:359]: Changed data type

File: src/libpgmoneta/configuration.c [hunk 6]
```diff
@@ -215,6 +215,7 @@ pgmoneta_read_configuration(void* shm, char* filename)
                   srv.cur_timeline = 1; // by default current timeline is 1
                   memset(srv.wal_shipping, 0, MAX_PATH);
                   srv.workers = -1;
+                  srv.network_max_rate=0;
```
Comment (jesperpedersen) [inline src/libpgmoneta/configuration.c:218]: Spaces

File: src/libpgmoneta/configuration.c [hunk 7]
```diff
@@ -1183,6 +1184,27 @@ pgmoneta_read_configuration(void* shm, char* filename)
...
+               else if (!strcmp(key, "network_max_rate"))
+               {
+                  if (!strcmp(section, "pgmoneta"))
+                  {
+                     if (as_ulong(value, &config->network_max_rate))
+                     {
+                        unknown = true;
+                     }
+                  }
+                  if (strlen(section) > 0)
```
Comment (jesperpedersen) [inline src/libpgmoneta/configuration.c:1196]: `else if`

File: src/libpgmoneta/configuration.c [hunk 8]
```diff
@@ -3011,6 +3033,8 @@ transfer_configuration(struct configuration* config, struct configuration* reloa
 
    config->workers = reload->workers;
 
+   config->network_max_rate = reload->network_max_rate;
```
Comment (jesperpedersen) [inline src/libpgmoneta/configuration.c:3036]: Move

File: src/libpgmoneta/configuration.c [hunk 9]
```diff
@@ -3044,6 +3068,7 @@ copy_server(struct server* dst, struct server* src)
    memcpy(&dst->current_wal_filename[0], &src->current_wal_filename[0], MISC_LENGTH);
    memcpy(&dst->current_wal_lsn[0], &src->current_wal_lsn[0], MISC_LENGTH);
    dst->workers = src->workers;
+   dst->network_max_rate=src->network_max_rate;
```
Comment (jesperpedersen) [inline src/libpgmoneta/configuration.c:3071]: Spaces

File: src/libpgmoneta/wf_backup.c [hunk 10]
```diff
@@ -125,6 +126,15 @@ basebackup_execute(int server, char* identifier, struct node* i_nodes, struct no
       }
    }
 
+   if (config->servers[server].network_max_rate || config->network_max_rate)
```
Comment (jesperpedersen) [inline src/libpgmoneta/wf_backup.c:129]: Provide a single function for this - could be in `utils.h`

**Review by jesperpedersen — CHANGES_REQUESTED** [review]

**Comment by jesperpedersen** [comment]
See the changes made by https://github.com/pgmoneta/pgmoneta/commit/5dd361f91c2c97323cead853db60f3498c6d6606

File: doc/CONFIGURATION.md (hunk 3 above)
Comment (shikharish) [reply doc/CONFIGURATION.md:95]: Is there a reason for `-1` to be default? I think `0` works fine too.

File: src/libpgmoneta/wf_backup.c (hunk 10 above)
Comment (jesperpedersen) [reply src/libpgmoneta/wf_backup.c:129]: It could be in `network.h` as well...

**Pull request merged**